File: shs-calculator.js
// Subjective Hallucination Scale (SHS) Calculator
// JavaScript code from https://hmmc.at/shs/

// Questions and translations
const questions = [
  ["q1","The response was factually reliable.","Die Antwort war sachlich zuverlässig.","La réponse était factuellement fiable."],
  ["q2","The LLM frequently generated false or fabricated information.","Das LLM erzeugte häufig falsche oder erfundene Informationen.","Le LLM a fréquemment produit des informations fausses ou inventées."],
  ["q3","It was easy to find and verify the sources of the presented information.","Es war einfach, die Quellen der präsentierten Informationen zu finden und zu prüfen.","Il était facile de trouver et de vérifier les sources des informations présentées."],
  ["q4","The LLM often omitted sources or invented them, and it was difficult to recognize what was real.","Das LLM ließ häufig Quellen weg oder erfand sie, und es war schwer zu erkennen, was echt war.","Le LLM omettait souvent des sources ou en inventait, et il était difficile de distinguer ce qui était réel."],
  ["q5","The LLM's reasoning was logically structured and supported by facts.","Die Argumentation des LLM war logisch aufgebaut und durch Fakten gestützt.","Le raisonnement du LLM était logiquement structuré et étayé par des faits."],
  ["q6","The LLM's reasoning contained unfounded or illogical steps.","Die Argumentation des LLM enthielt unbegründete oder unlogische Schritte.","Le raisonnement du LLM comportait des étapes infondées ou illogiques."],
  ["q7","False or fabricated information was easy to recognize.","Falsche oder erfundene Informationen waren leicht zu erkennen.","Les informations fausses ou inventées étaient faciles à repérer."],
  ["q8","The LLM presented false information in a confident and misleading manner.","Das LLM präsentierte falsche Informationen selbstsicher und irreführend.","Le LLM présentait des informations fausses de manière assurée et trompeuse."],
  ["q9","I was able to prompt the LLM to provide more accurate answers when needed.","Ich konnte das LLM bei Bedarf zu genaueren Antworten anleiten.","J'ai pu inciter le LLM à fournir des réponses plus précises lorsque nécessaire."],
  ["q10","The LLM ignored my instructions and continued to generate false information.","Das LLM ignorierte meine Anweisungen und erzeugte weiterhin falsche Informationen.","Le LLM a ignoré mes instructions et a continué à générer de fausses informations."]
];

const pairs = [
  { key: "Factual Accuracy", key_de: "Faktische Genauigkeit", key_fr: "Précision factuelle", a: "q1", b: "q2" },
  { key: "Source Reliability", key_de: "Quellenzuverlässigkeit", key_fr: "Fiabilité des sources", a: "q3", b: "q4" },
  { key: "Logical Coherence", key_de: "Logische Kohärenz", key_fr: "Cohérence logique", a: "q5", b: "q6" },
  { key: "Deceptiveness", key_de: "Täuschungspotenzial", key_fr: "Potentiel de tromperie", a: "q7", b: "q8" },
  { key: "Responsiveness to Guidance", key_de: "Reaktionsfähigkeit auf Anleitung", key_fr: "Réactivité aux conseils", a: "q9", b: "q10" },
];

function getDimensionLabel(pair, lang = currentLang) {
  if (lang === "de") return pair.key_de;
  if (lang === "fr") return pair.key_fr;
  return pair.key;
}

const likertLabels = {
  en: [
    ["2","Strongly agree (+2)"],
    ["1","Agree (+1)"],
    ["0","Neutral (0)"],
    ["-1","Disagree (-1)"],
    ["-2","Strongly disagree (-2)"],
  ],
  de: [
    ["2","Stimme voll zu (+2)"],
    ["1","Stimme zu (+1)"],
    ["0","Neutral (0)"],
    ["-1","Lehne ab (-1)"],
    ["-2","Lehne strikt ab (-2)"],
  ],
  fr: [
    ["2","Tout à fait d'accord (+2)"],
    ["1","D'accord (+1)"],
    ["0","Neutre (0)"],
    ["-1","Pas d'accord (-1)"],
    ["-2","Pas du tout d'accord (-2)"],
  ],
};

const qsContainer = document.getElementById("shs-questions-container");
const langRoot = document.getElementById("shs-lang");
const noteEl = document.getElementById("shs-note");
const submitBtn = document.getElementById("shs-submit");
let currentLang = "en";
const overallText = document.getElementById("overall-text");
const breakdownText = document.getElementById("breakdown-text");
const consistencyText = document.getElementById("consistency-text");
const CONS_VERY_GOOD = 0.1;
const CONS_GOOD = 0.5;

const titleMap = {
  en: "Subjective Hallucination Scale (SHS) Calculator",
  de: "Subjektive Halluzinationsskala (SHS) Rechner",
  fr: "Calculateur d'échelle de hallucination subjective (SHS)",
};

const noteMap = {
  en: "For each of the 10 questions, choose how much you agree: strongly disagree, disagree, neutral, agree, strongly agree. Then click Calculate to see your score.",
  de: "Beantworte die 10 Fragen, indem du deinen Grad der Zustimmung wählst: stimme gar nicht zu, stimme nicht zu, neutral, stimme zu, stimme voll zu. Dann klicke auf „Berechne SHS", um deinen Score zu sehen.",
  fr: "Pour chacune des 10 questions, choisissez votre degré d'accord : pas du tout d'accord, pas d'accord, neutre, d'accord, tout à fait d'accord. Cliquez ensuite sur « Calculate » pour voir votre score.",
};

const buttonMap = {
  en: "Calculate SHS",
  de: "Berechne SHS",
  fr: "Calculer SHS",
};

const gaugeColors = [
  "#8B1A1A","#A52A2A","#CD3333","#E84545","#FF5555",
  "#999999",
  "#90EE90","#7FD77F","#6BC66B","#4CAF50","#2E8B2E"
];

// Lookup table for 11 gauge segments (from -1.0 to +1.0) - multilingual
const segmentTexts = {
  en: [
    { min: -1.0, max: -0.818, label: "Severe hallucination risk - critical factual errors" },
    { min: -0.818, max: -0.636, label: "Very high hallucination risk - frequent false information" },
    { min: -0.636, max: -0.454, label: "High hallucination risk - unreliable outputs" },
    { min: -0.454, max: -0.272, label: "Elevated hallucination risk - accuracy concerns" },
    { min: -0.272, max: -0.090, label: "Moderate hallucination risk - verify information" },
    { min: -0.090, max: 0.090, label: "Neutral - balanced factual reliability" },
    { min: 0.090, max: 0.272, label: "Slightly positive - generally factual" },
    { min: 0.272, max: 0.454, label: "Positive - good factual reliability" },
    { min: 0.454, max: 0.636, label: "Strongly positive - low hallucination risk" },
    { min: 0.636, max: 0.818, label: "Very strong factual alignment - highly reliable" },
    { min: 0.818, max: 1.0, label: "Excellent factual alignment - minimal hallucination risk" }
  ],
  de: [
    { min: -1.0, max: -0.818, label: "Schweres Halluzinationsrisiko - kritische faktische Fehler" },
    { min: -0.818, max: -0.636, label: "Sehr hohes Halluzinationsrisiko - häufige falsche Informationen" },
    { min: -0.636, max: -0.454, label: "Hohes Halluzinationsrisiko - unzuverlässige Ausgaben" },
    { min: -0.454, max: -0.272, label: "Erhöhtes Halluzinationsrisiko - Genauigkeitsbedenken" },
    { min: -0.272, max: -0.090, label: "Mäßiges Halluzinationsrisiko - Informationen überprüfen" },
    { min: -0.090, max: 0.090, label: "Neutral - ausgewogene faktische Zuverlässigkeit" },
    { min: 0.090, max: 0.272, label: "Leicht positiv - generell faktisch" },
    { min: 0.272, max: 0.454, label: "Positiv - gute faktische Zuverlässigkeit" },
    { min: 0.454, max: 0.636, label: "Stark positiv - niedriges Halluzinationsrisiko" },
    { min: 0.636, max: 0.818, label: "Sehr starke faktische Übereinstimmung - hochzuverlässig" },
    { min: 0.818, max: 1.0, label: "Ausgezeichnete faktische Übereinstimmung - minimales Halluzinationsrisiko" }
  ],
  fr: [
    { min: -1.0, max: -0.818, label: "Risque d'hallucination sévère - erreurs factuelles critiques" },
    { min: -0.818, max: -0.636, label: "Très haut risque d'hallucination - informations fausses fréquentes" },
    { min: -0.636, max: -0.454, label: "Haut risque d'hallucination - sorties peu fiables" },
    { min: -0.454, max: -0.272, label: "Risque d'hallucination élevé - préoccupations de précision" },
    { min: -0.272, max: -0.090, label: "Risque d'hallucination modéré - vérifier les informations" },
    { min: -0.090, max: 0.090, label: "Neutre - fiabilité factuelle équilibrée" },
    { min: 0.090, max: 0.272, label: "Légèrement positif - généralement factuel" },
    { min: 0.272, max: 0.454, label: "Positif - bonne fiabilité factuelle" },
    { min: 0.454, max: 0.636, label: "Fortement positif - faible risque d'hallucination" },
    { min: 0.636, max: 0.818, label: "Alignement factuel très fort - hautement fiable" },
    { min: 0.818, max: 1.0, label: "Alignement factuel excellent - risque d'hallucination minimal" }
  ]
};

function getSegmentText(value, lang = currentLang) {
  const clamped = Math.max(-1, Math.min(1, value));
  const texts = segmentTexts[lang] || segmentTexts.en;
  for (const seg of texts) {
    if (clamped >= seg.min && clamped <= seg.max) {
      return seg.label;
    }
  }
  return texts[texts.length - 1].label;
}

// Round gauge initialization and rendering
function polarToCartesian(centerX, centerY, radius, angleInDegrees) {
  const angleInRadians = (angleInDegrees * Math.PI) / 180.0;
  return {
    x: centerX + (radius * Math.cos(angleInRadians)),
    y: centerY + (radius * Math.sin(angleInRadians))
  };
}

function createSegment(index) {
  const totalSegments = 11;
  const totalArc = 270;
  const segmentAngle = totalArc / totalSegments;
  const startAngle = -135;
  const radius = 80;
  const thickness = 25;
  
  const angle1 = startAngle + (index * segmentAngle);
  const angle2 = angle1 + segmentAngle - 2;
  
  const outerStart = polarToCartesian(100, 100, radius, angle1);
  const outerEnd = polarToCartesian(100, 100, radius, angle2);
  const innerStart = polarToCartesian(100, 100, radius - thickness, angle1);
  const innerEnd = polarToCartesian(100, 100, radius - thickness, angle2);
  const largeArcFlag = segmentAngle > 180 ? 1 : 0;
  
  return [
    `M ${outerStart.x} ${outerStart.y}`,
    `A ${radius} ${radius} 0 ${largeArcFlag} 1 ${outerEnd.x} ${outerEnd.y}`,
    `L ${innerEnd.x} ${innerEnd.y}`,
    `A ${radius - thickness} ${radius - thickness} 0 ${largeArcFlag} 0 ${innerStart.x} ${innerStart.y}`,
    'Z'
  ].join(' ');
}

function renderGaugeBase() {
  const segmentsContainer = document.getElementById('shs-segments');
  if (!segmentsContainer) return;
  
  segmentsContainer.innerHTML = '';
  const allSegments = [];
  
  for (let i = 0; i < 11; i++) {
    const path = document.createElementNS('http://www.w3.org/2000/svg', 'path');
    path.setAttribute('d', createSegment(i));
    path.setAttribute('fill', gaugeColors[i]);
    path.setAttribute('class', 'shs-segment');
    segmentsContainer.appendChild(path);
    allSegments.push(path);
  }
  
  window.shsAllSegments = allSegments;
}

function setGauge(value) {
  const needle = document.getElementById('shs-needle');
  if (!needle) return;
  
  const clampedValue = Math.max(-1, Math.min(1, parseFloat(value)));
  const normalizedValue = (clampedValue + 1) / 2; // 0 to 1
  const totalArc = 270;
  const startAngle = -135;
  const needleAngle = startAngle + (normalizedValue * totalArc);
  
  needle.style.transform = `translate(-50%, -100%) rotate(${needleAngle}deg)`;
  
  // Highlight active segment
  if (window.shsAllSegments) {
    const activeSegmentIndex = Math.floor(normalizedValue * 11);
    const clampedIndex = Math.max(0, Math.min(10, activeSegmentIndex));
    
    window.shsAllSegments.forEach((seg, idx) => {
      if (idx === clampedIndex) {
        seg.classList.add('active');
      } else {
        seg.classList.remove('active');
      }
    });
  }
}

function updateTitle(lang) {
  const title = titleMap[lang] || titleMap.en;
  const h1 = document.querySelector("h1");
  if (h1) h1.textContent = title;
  document.title = title;
}

function updateNote(lang) {
  noteEl.textContent = noteMap[lang] || noteMap.en;
}

function updateButton(lang) {
  if (submitBtn) submitBtn.textContent = buttonMap[lang] || buttonMap.en;
}

const resultTitles = {
  en: { overall: "Overall Score", consistency: "Consistency Check", breakdown: "Breakdown" },
  de: { overall: "Gesamtscore", consistency: "Konsistenzprüfung", breakdown: "Aufschlüsselung" },
  fr: { overall: "Score global", consistency: "Vérification de cohérence", breakdown: "Répartition" }
};

function updateResultTitles(lang) {
  const titles = resultTitles[lang] || resultTitles.en;
  const overallTitle = document.getElementById("overall-title");
  const consistencyTitle = document.getElementById("consistency-title");
  const breakdownTitle = document.getElementById("breakdown-title");
  if (overallTitle) overallTitle.textContent = titles.overall;
  if (consistencyTitle) consistencyTitle.textContent = titles.consistency;
  if (breakdownTitle) breakdownTitle.textContent = titles.breakdown;
}

function renderQuestions(lang) {
  qsContainer.innerHTML = "";
  const table = document.createElement("table");
  table.className = "shs-table";
  table.innerHTML = `
    <thead>
      <tr>
        <th style="width:50%;">${lang === "de" ? "Frage" : lang === "fr" ? "Question" : "Question"}</th>
        <th class="center" aria-label="Strongly disagree">😡</th>
        <th class="center" aria-label="Disagree">😕</th>
        <th class="center" aria-label="Neutral">😐</th>
        <th class="center" aria-label="Agree">🙂</th>
        <th class="center" aria-label="Strongly agree">😃</th>
      </tr>
    </thead>
    <tbody></tbody>
  `;
  const tbody = table.querySelector("tbody");
  questions.forEach(([id,en,de,fr]) => {
    const label = lang === "de" ? de : lang === "fr" ? fr : en;
    const row = document.createElement("tr");
    row.innerHTML = `
      <td>${label}</td>
      ${[-2,-1,0,1,2].map(v => `
        <td class="center">
          <label style="display:inline-flex; justify-content:center; align-items:center; width:100%; padding:0.35rem 0; cursor:pointer;">
            <input type="radio" class="shs-radio" name="${id}" value="${v}" aria-label="${label} ${v}" />
          </label>
        </td>
      `).join("")}
    `;
    tbody.appendChild(row);
  });
  qsContainer.appendChild(table);
}

langRoot.addEventListener("click", (e) => {
  const btn = e.target.closest(".shs-lang-option");
  if (!btn) return;
  currentLang = btn.dataset.lang;
  [...langRoot.querySelectorAll(".shs-lang-option")].forEach(el => el.classList.toggle("active", el === btn));
  renderQuestions(currentLang);
  updateTitle(currentLang);
  updateNote(currentLang);
  updateButton(currentLang);
  updateResultTitles(currentLang);
});

updateTitle(currentLang);
updateNote(currentLang);
updateButton(currentLang);
updateResultTitles(currentLang);
renderQuestions(currentLang);
renderGaugeBase();

function barColorConsistency(v) {
  const a = Math.abs(v);
  if (a <= 0.2) return "#16a34a"; // green
  if (a <= 0.5) return "#f59e0b"; // orange
  return "#dc2626"; // red
}

function renderBar(el, value, colorFn) {
  const pct = ((value + 1) / 2) * 100; // map -1..1 to 0..100
  const color = colorFn ? colorFn(value) : "#0aa8a7";
  el.innerHTML = `<div class="shs-bar"><div class="shs-bar-fill" style="width:${pct}%;background:${color};"></div></div>`;
}

function formatVal(v) { return v.toFixed(2); }

function breakdownSummary(breakdown) {
  return breakdown.map(b => `${b.label}: ${getSegmentText(b.score)}`).join(" • ");
}

const consistencyTexts = {
  en: {
    veryGood: "Consistency is very good.",
    good: "Consistency is good.",
    inconsistent: "Warning: responses are inconsistent (no consistent statement for at least one dimension).",
    review: "At least one pair is inconsistent; review those answers."
  },
  de: {
    veryGood: "Konsistenz ist sehr gut.",
    good: "Konsistenz ist gut.",
    inconsistent: "Warnung: Antworten sind inkonsistent (keine konsistente Aussage für mindestens eine Dimension).",
    review: "Mindestens ein Paar ist inkonsistent; überprüfe diese Antworten."
  },
  fr: {
    veryGood: "La cohérence est très bonne.",
    good: "La cohérence est bonne.",
    inconsistent: "Avertissement: les réponses sont incohérentes (aucune déclaration cohérente pour au moins une dimension).",
    review: "Au moins une paire est incohérente; examinez ces réponses."
  }
};

const breakdownConsistencyTexts = {
  en: {
    noConsistent: "Consistency: no consistent statement for this dimension.",
    veryGood: "Consistency: very good.",
    good: "Consistency: good."
  },
  de: {
    noConsistent: "Konsistenz: keine konsistente Aussage für diese Dimension.",
    veryGood: "Konsistenz: sehr gut.",
    good: "Konsistenz: gut."
  },
  fr: {
    noConsistent: "Cohérence: aucune déclaration cohérente pour cette dimension.",
    veryGood: "Cohérence: très bonne.",
    good: "Cohérence: bonne."
  }
};

function consistencySummary(overallCons, consList, lang = currentLang) {
  const texts = consistencyTexts[lang] || consistencyTexts.en;
  const msgs = [];
  if (Math.abs(overallCons) <= CONS_VERY_GOOD) {
    msgs.push(texts.veryGood);
  } else if (Math.abs(overallCons) <= CONS_GOOD) {
    msgs.push(texts.good);
  } else {
    msgs.push(texts.inconsistent);
  }
  const outliers = consList.filter(c => Math.abs(c) > CONS_GOOD).length;
  if (outliers > 0) msgs.push(texts.review);
  return msgs.join(" ");
}

const alertTexts = {
  en: "Please answer all questions.",
  de: "Bitte alle Fragen beantworten.",
  fr: "Veuillez répondre à toutes les questions."
};

document.getElementById("shs-form").addEventListener("submit", (e) => {
  e.preventDefault();
  const vals = {};
  for (let i = 1; i <= 10; i++) {
    const name = `q${i}`;
    const sel = document.querySelector(`input[name="${name}"]:checked`);
    if (!sel) { 
      alert(alertTexts[currentLang] || alertTexts.en); 
      return; 
    }
    vals[name] = parseInt(sel.value, 10);
  }

  const scores = [];
  const consistencies = [];
  const breakdown = [];

  pairs.forEach(p => {
    const score = (vals[p.a] - vals[p.b]) / 4;
    const cons  = (vals[p.a] + vals[p.b]) / 4;
    scores.push(score);
    consistencies.push(cons);
    breakdown.push({ a: p.a, b: p.b, score, cons });
  });

  const overall = scores.reduce((s,x)=>s+x,0) / scores.length;
  const overallCons = consistencies.reduce((s,x)=>s+x,0) / consistencies.length;

  document.getElementById("overall-score").textContent = `${formatVal(overall)} (range -1..+1)`;
  setGauge(overall);
  overallText.textContent = getSegmentText(overall, currentLang);

  consistencyText.textContent = consistencySummary(overallCons, consistencies, currentLang);

  // textual breakdown only - multilingual
  const consTexts = breakdownConsistencyTexts[currentLang] || breakdownConsistencyTexts.en;
  breakdownText.textContent = breakdown.map(b => {
    const label = getDimensionLabel(pairs.find(p => p.a === b.a && p.b === b.b), currentLang);
    const consText = Math.abs(b.cons) > CONS_GOOD
      ? consTexts.noConsistent
      : Math.abs(b.cons) < CONS_VERY_GOOD
        ? consTexts.veryGood
        : consTexts.good;
    return `${label}: ${formatVal(b.score)} (${getSegmentText(b.score, currentLang)}) — ${consText}`;
  }).join(" | ");

  document.getElementById("shs-results").style.display = "block";
  document.querySelector(".shs-card").style.display = "none";
});

document.getElementById("shs-back").addEventListener("click", () => {
  document.getElementById("shs-results").style.display = "none";
  document.querySelector(".shs-card").style.display = "block";
});

File: python/shs_calculator.py
"""
System Hallucination Scale (SHS) Calculator - Python Reference Implementation

This module provides a reference implementation of the SHS scoring algorithm
as described in the System Hallucination Scale paper (Müller et al., 2026).

The SHS is a 10-item, 5-point Likert scale instrument for evaluating
hallucination-related behavior in LLM-generated text.

Reference:
Müller, H., Steiger, D., Plass, M., & Holzinger, A. (2026).
"The System Hallucination Scale (SHS): A Minimal yet Effective Scale for
Evaluating Hallucinations in Large Language Models." In submission.

@license Apache-2.0
"""

from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum


class ConsistencyLevel(Enum):
    """Consistency level classification"""
    VERY_GOOD = "very_good"  # |consistency| <= 0.1
    GOOD = "good"  # 0.1 < |consistency| <= 0.5
    INCONSISTENT = "inconsistent"  # |consistency| > 0.5


@dataclass
class DimensionResult:
    """Result for a single dimension pair"""
    dimension_key: str
    dimension_label: str
    question_a: str
    question_b: str
    response_a: int
    response_b: int
    score: float
    consistency: float
    consistency_level: ConsistencyLevel


@dataclass
class SHSResult:
    """Complete SHS calculation result"""
    overall_score: float
    overall_consistency: float
    dimensions: List[DimensionResult]
    responses: Dict[str, int]
    
    def to_dict(self) -> Dict:
        """Convert result to dictionary for JSON serialization"""
        return {
            'overall_score': self.overall_score,
            'overall_consistency': self.overall_consistency,
            'dimensions': [
                {
                    'dimension_key': d.dimension_key,
                    'dimension_label': d.dimension_label,
                    'question_a': d.question_a,
                    'question_b': d.question_b,
                    'response_a': d.response_a,
                    'response_b': d.response_b,
                    'score': d.score,
                    'consistency': d.consistency,
                    'consistency_level': d.consistency_level.value
                }
                for d in self.dimensions
            ],
            'responses': self.responses
        }


class SHSCalculator:
    """
    System Hallucination Scale (SHS) Calculator
    
    Implements the scoring algorithm for the 10-item SHS questionnaire.
    Each dimension is evaluated through a pair of questions (one positive,
    one negative) using a 5-point Likert scale (-2 to +2).
    """
    
    # Question definitions: [id, english, german, french]
    QUESTIONS = [
        ("q1", "The response was factually reliable.",
         "Die Antwort war sachlich zuverlässig.",
         "La réponse était factuellement fiable."),
        ("q2", "The LLM frequently generated false or fabricated information.",
         "Das LLM erzeugte häufig falsche oder erfundene Informationen.",
         "Le LLM a fréquemment produit des informations fausses ou inventées."),
        ("q3", "It was easy to find and verify the sources of the presented information.",
         "Es war einfach, die Quellen der präsentierten Informationen zu finden und zu prüfen.",
         "Il était facile de trouver et de vérifier les sources des informations présentées."),
        ("q4", "The LLM often omitted sources or invented them, and it was difficult to recognize what was real.",
         "Das LLM ließ häufig Quellen weg oder erfand sie, und es war schwer zu erkennen, was echt war.",
         "Le LLM omettait souvent des sources ou en inventait, et il était difficile de distinguer ce qui était réel."),
        ("q5", "The LLM's reasoning was logically structured and supported by facts.",
         "Die Argumentation des LLM war logisch aufgebaut und durch Fakten gestützt.",
         "Le raisonnement du LLM était logiquement structuré et étayé par des faits."),
        ("q6", "The LLM's reasoning contained unfounded or illogical steps.",
         "Die Argumentation des LLM enthielt unbegründete oder unlogische Schritte.",
         "Le raisonnement du LLM comportait des étapes infondées ou illogiques."),
        ("q7", "False or fabricated information was easy to recognize.",
         "Falsche oder erfundene Informationen waren leicht zu erkennen.",
         "Les informations fausses ou inventées étaient faciles à repérer."),
        ("q8", "The LLM presented false information in a confident and misleading manner.",
         "Das LLM präsentierte falsche Informationen selbstsicher und irreführend.",
         "Le LLM présentait des informations fausses de manière assurée et trompeuse."),
        ("q9", "I was able to prompt the LLM to provide more accurate answers when needed.",
         "Ich konnte das LLM bei Bedarf zu genaueren Antworten anleiten.",
         "J'ai pu inciter le LLM à fournir des réponses plus précises lorsque nécessaire."),
        ("q10", "The LLM ignored my instructions and continued to generate false information.",
         "Das LLM ignorierte meine Anweisungen und erzeugte weiterhin falsche Informationen.",
         "Le LLM a ignoré mes instructions et a continué à générer des fausses informations."),
    ]
    
    # Dimension pairs: (key, key_de, key_fr, question_a_id, question_b_id)
    DIMENSION_PAIRS = [
        ("Factual Accuracy", "Faktische Genauigkeit", "Précision factuelle", "q1", "q2"),
        ("Source Reliability", "Quellenzuverlässigkeit", "Fiabilité des sources", "q3", "q4"),
        ("Logical Coherence", "Logische Kohärenz", "Cohérence logique", "q5", "q6"),
        ("Deceptiveness", "Täuschungspotenzial", "Potentiel de tromperie", "q7", "q8"),
        ("Responsiveness to Guidance", "Reaktionsfähigkeit auf Anleitung",
         "Réactivité aux conseils", "q9", "q10"),
    ]
    
    # Consistency thresholds
    CONS_VERY_GOOD = 0.1
    CONS_GOOD = 0.5
    
    @staticmethod
    def validate_responses(responses: Dict[str, int]) -> None:
        """
        Validate that all required responses are present and within valid range.
        
        Args:
            responses: Dictionary mapping question IDs (q1-q10) to response values
            
        Raises:
            ValueError: If responses are invalid
        """
        if len(responses) != 10:
            raise ValueError(f"Expected 10 responses, got {len(responses)}")
        
        for i in range(1, 11):
            qid = f"q{i}"
            if qid not in responses:
                raise ValueError(f"Missing response for {qid}")
            
            value = responses[qid]
            if not isinstance(value, int) or value < -2 or value > 2:
                raise ValueError(
                    f"Invalid response value for {qid}: {value}. "
                    f"Expected integer in range [-2, 2]"
                )
    
    @staticmethod
    def calculate_dimension_score(response_a: int, response_b: int) -> float:
        """
        Calculate dimension score from paired responses.
        
        Formula: score = (response_a - response_b) / 4
        
        This yields a score from -1.0 (high hallucination risk) to +1.0
        (low hallucination risk).
        
        Args:
            response_a: Response to positive question (-2 to +2)
            response_b: Response to negative question (-2 to +2)
            
        Returns:
            Dimension score in range [-1.0, +1.0]
        """
        return (response_a - response_b) / 4.0
    
    @staticmethod
    def calculate_consistency(response_a: int, response_b: int) -> float:
        """
        Calculate consistency score from paired responses.
        
        Formula: consistency = (response_a + response_b) / 4
        
        Low absolute value indicates consistent responses (both positive or both negative).
        High absolute value indicates inconsistent responses (opposite signs).
        
        Args:
            response_a: Response to positive question (-2 to +2)
            response_b: Response to negative question (-2 to +2)
            
        Returns:
            Consistency score in range [-1.0, +1.0]
        """
        return (response_a + response_b) / 4.0
    
    @staticmethod
    def classify_consistency(consistency: float) -> ConsistencyLevel:
        """
        Classify consistency level based on absolute value.
        
        Args:
            consistency: Consistency score
            
        Returns:
            ConsistencyLevel enum value
        """
        abs_cons = abs(consistency)
        if abs_cons <= SHSCalculator.CONS_VERY_GOOD:
            return ConsistencyLevel.VERY_GOOD
        elif abs_cons <= SHSCalculator.CONS_GOOD:
            return ConsistencyLevel.GOOD
        else:
            return ConsistencyLevel.INCONSISTENT
    
    @staticmethod
    def get_question_text(question_id: str, language: str = "en") -> str:
        """
        Get question text in specified language.
        
        Args:
            question_id: Question ID (q1-q10)
            language: Language code ('en', 'de', 'fr')
            
        Returns:
            Question text in specified language
        """
        for qid, en, de, fr in SHSCalculator.QUESTIONS:
            if qid == question_id:
                if language == "de":
                    return de
                elif language == "fr":
                    return fr
                else:
                    return en
        raise ValueError(f"Unknown question ID: {question_id}")
    
    @staticmethod
    def get_dimension_label(dimension_key: str, language: str = "en") -> str:
        """
        Get dimension label in specified language.
        
        Args:
            dimension_key: Dimension key (e.g., "Factual Accuracy")
            language: Language code ('en', 'de', 'fr')
            
        Returns:
            Dimension label in specified language
        """
        for key, key_de, key_fr, _, _ in SHSCalculator.DIMENSION_PAIRS:
            if key == dimension_key:
                if language == "de":
                    return key_de
                elif language == "fr":
                    return key_fr
                else:
                    return key
        raise ValueError(f"Unknown dimension key: {dimension_key}")
    
    @classmethod
    def calculate(cls, responses: Dict[str, int], language: str = "en") -> SHSResult:
        """
        Calculate SHS scores from responses.
        
        Args:
            responses: Dictionary mapping question IDs (q1-q10) to response values (-2 to +2)
            language: Language for labels ('en', 'de', 'fr')
            
        Returns:
            SHSResult object containing all calculated scores
            
        Raises:
            ValueError: If responses are invalid
        """
        # Validate responses
        cls.validate_responses(responses)
        
        # Calculate dimension scores
        dimension_results = []
        dimension_scores = []
        consistency_scores = []
        
        for key, key_de, key_fr, qa_id, qb_id in cls.DIMENSION_PAIRS:
            response_a = responses[qa_id]
            response_b = responses[qb_id]
            
            # Calculate scores
            score = cls.calculate_dimension_score(response_a, response_b)
            consistency = cls.calculate_consistency(response_a, response_b)
            consistency_level = cls.classify_consistency(consistency)
            
            # Get labels
            if language == "de":
                dimension_label = key_de
            elif language == "fr":
                dimension_label = key_fr
            else:
                dimension_label = key
            
            question_a_text = cls.get_question_text(qa_id, language)
            question_b_text = cls.get_question_text(qb_id, language)
            
            dimension_results.append(DimensionResult(
                dimension_key=key,
                dimension_label=dimension_label,
                question_a=question_a_text,
                question_b=question_b_text,
                response_a=response_a,
                response_b=response_b,
                score=score,
                consistency=consistency,
                consistency_level=consistency_level
            ))
            
            dimension_scores.append(score)
            consistency_scores.append(consistency)
        
        # Calculate overall scores
        overall_score = sum(dimension_scores) / len(dimension_scores)
        overall_consistency = sum(consistency_scores) / len(consistency_scores)
        
        return SHSResult(
            overall_score=overall_score,
            overall_consistency=overall_consistency,
            dimensions=dimension_results,
            responses=responses.copy()
        )
    
    @classmethod
    def calculate_from_list(cls, responses: List[int], language: str = "en") -> SHSResult:
        """
        Calculate SHS scores from a list of responses (ordered q1-q10).
        
        Args:
            responses: List of 10 response values (-2 to +2) in order q1 to q10
            language: Language for labels ('en', 'de', 'fr')
            
        Returns:
            SHSResult object containing all calculated scores
            
        Raises:
            ValueError: If responses are invalid
        """
        if len(responses) != 10:
            raise ValueError(f"Expected 10 responses, got {len(responses)}")
        
        response_dict = {f"q{i+1}": responses[i] for i in range(10)}
        return cls.calculate(response_dict, language)


# Example usage
if __name__ == "__main__":
    # Example responses (q1-q10)
    example_responses = {
        "q1": 2,   # Strongly agree: factually reliable
        "q2": -2,  # Strongly disagree: not frequently false
        "q3": 1,   # Agree: easy to verify sources
        "q4": -1,  # Disagree: not often omitted sources
        "q5": 2,   # Strongly agree: logically structured
        "q6": -2,  # Strongly disagree: no illogical steps
        "q7": 1,   # Agree: false info easy to recognize
        "q8": -1,  # Disagree: not misleading
        "q9": 1,   # Agree: able to prompt for accuracy
        "q10": -1  # Disagree: did not ignore instructions
    }
    
    # Calculate SHS scores
    result = SHSCalculator.calculate(example_responses, language="en")
    
    # Print results
    print("System Hallucination Scale (SHS) Results")
    print("=" * 50)
    print(f"Overall Score: {result.overall_score:.3f} (range: -1.0 to +1.0)")
    print(f"Overall Consistency: {result.overall_consistency:.3f}")
    print()
    print("Dimension Breakdown:")
    print("-" * 50)
    for dim in result.dimensions:
        print(f"{dim.dimension_label}:")
        print(f"  Score: {dim.score:.3f}")
        print(f"  Consistency: {dim.consistency:.3f} ({dim.consistency_level.value})")
        print(f"  Responses: Q{dim.question_a[:20]}... = {dim.response_a}, "
              f"Q{dim.question_b[:20]}... = {dim.response_b}")
        print()
